usersession/userd/launcher.go

```diff
@@ -123,6 +131,105 @@ func (s *Launcher) OpenURL(addr string, sender dbus.Sender) *dbus.Error {
 	return nil
 }
 
+// OpenDesktopEntry implements the 'OpenDesktopEntry' method of the 'io.snapcraft.Launcher'
+// DBus interface. Before the provided desktop_file is parsed it is validated against a list
+// of allowed locations.
+func (s *Launcher) OpenDesktopEntry(desktop_file_id string, sender dbus.Sender) *dbus.Error {
+
+	return s.OpenDesktopEntryEnv(desktop_file_id, []string{}, sender)
+}
+
+// OpenDesktopEntryEnv implements the 'OpenDesktopEntryEnv' method of the 'io.snapcraft.Launcher'
+// DBus interface.
+func (s *Launcher) OpenDesktopEntryEnv(desktop_file_id string, env []string, sender dbus.Sender) *dbus.Error {
+	exec_command, err := s.readExecCommandFromDesktopFile(s.desktopFileIdToFilename(desktop_file_id))
+
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	// Passing exec variables between confined snaps raises unanswered questions and they are not required
+	// for the simple cases.  For now, we don't have support for passing them in the dbus API and truncate
+	// the exec command at the first exec variable.
+	// https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#exec-variables
+	exec_command = strings.SplitN(exec_command, "%", 2)[0]
+
+	args := strings.Split(exec_command, " ")
+	cmd := exec.Command(args[0], args[1:]...)
+
+	cmd.Env = os.Environ()
+	for _, e := range env {
+		cmd.Env = append(cmd.Env, e)
+	}
```

Comment:
Should there be a whitelist of variables the caller is allowed to set?  Also, does it make sense to set variables on a "per exec" basis rather than once for the session?  If you're using it for things like the Wayland display socket, then surely you'd want this to also cover processes not directly launched by the shell (e.g. D-Bus services).

Maybe a second API that can do `systemctl --user set-environment` or `dbus-update-activation-environment` would be appropriate?

Reply:
> Should there be a whitelist of variables the caller is allowed to set?

I know that I currently only want `WAYLAND_DISPLAY`, but others (such as `XDG_CURRENT_DESKTOP`, `XDG_SESSION_DESKTOP` and `XDG_SESSION_TYPE`) make sense.

I can do that.

>  Also, does it make sense to set variables on a "per exec" basis rather than once for the session?

Setting the session environment makes sense when the confined server is THE Wayland server, but doesn't cover potentially important cases when it is not.

Running a confined desktop in a window on "Ubuntu Wayland" (or other desktop) is entirely feasible and potentially useful. In this scenario setting the session `WAYLAND_DISPLAY` would mean any apps launched will connect to the confined desktop.

Reply:
This seems to be focusing on a fringe use case (running two desktops within the same user session) at the expense of a common use case (processes on the desktop started via other means, such as D-Bus activation).